### PR TITLE
Reference `PL_GLOBAL_SEED` in `pl_worker_init_function()` to correctly seed dataloader workers

### PR DESCRIPTION
**Pull request by amorehead**
## What does this PR do?

Fixes an issue where the value for `PL_GLOBAL_SEED` set by `seed_everything()` would not be referenced in `pl_worker_init_function()`. This could cause multi-process dataloader workers to not change their random state based on a user's desired seed set via `seed_everything()` before training. In other words, if users were relying on `seed_everything()` to change e.g., the order of the indices sampled by each dataloader worker, this would not work until now.

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

**Comment by lantiga** [comment]
Nice catch! Can you please add a test that verifies the desired behavior so we know it won't regress in the future?

**Review by Borda — APPROVED** [review]

**Comment by adosar** [comment]
@amorehead Does this mean that previously `seed_everything` didn't work as expected?

**Comment by amorehead** [comment]
@adosar, if your multiprocessing dataloaders (plural) use NumPy's random module for index sampling, then yes, until this PR was merged in Lightning was not handling random seed definitions correctly.

**Comment by adosar** [comment]
@amorehead If I understand correctly, this affect any dataset with a `np.random.*` call under `__getitem__`?

**Comment by amorehead** [comment]
@adosar, yes, any `np.random.*` calls under `__getitem__` (if you used multiple dataloader workers) would not be seeded properly until now.

**Comment by adosar** [comment]
@amorehead How does `pl_worker_init_function` accounts for different seeds for different epochs? In the following snippet:

```python
...
process_seed = torch.initial_seed()   # Returns what was set from the user, e.g. 42
base_seed = process_seed - worker_id
... 
if _NUMPY_AVAILABLE:
    import numpy as np
    
    ss = np.random.SeedSequence([base_seed, worker_id, global_rank])
    np_rng_seed = ss.generate_state(4)
    
    np.random.seed(np_rng_seed)
```
the generated seed is dependent on `process_id` and `worker_id`, which can be identical for different epochs.
